stag-library-compiler/src/main/java/com/vimeo/stag/processor/generators/TypeAdapterGenerator.java

```diff
@@ -604,13 +682,40 @@ private AdapterFieldInfo addAdapterFields(@Nullable StagGenerator.GenericClassIn
                                               @NotNull Map<TypeVariable, String> typeVarsMap,
                                               @NotNull StagGenerator stagGenerator) {
 
-        HashSet<TypeMirror> typeSet = new HashSet<>(memberVariables.values());
-        AdapterFieldInfo result = new AdapterFieldInfo(typeSet.size());
+        AdapterFieldInfo result = new AdapterFieldInfo(memberVariables.size());
         boolean hasUnknownGenericField =
                 genericClassInfo != null && genericClassInfo.mHasUnknownVarTypeFields;
-        for (TypeMirror fieldType : typeSet) {
-            String adapterAccessor;
-            if (hasUnknownGenericField && TypeUtils.containsTypeVarParams(fieldType)) {
+        for (Map.Entry<Element, TypeMirror> entry : memberVariables.entrySet()) {
+            TypeMirror fieldType = entry.getValue();
+            JsonAdapter annotation = entry.getKey().getAnnotation(JsonAdapter.class);
+            String adapterAccessor = null;
+            if(null != annotation) {
+                TypeMirror jsonAdapterType = null;
+                Element jsonAdapterTypeElement = null;
+                // Using this trick to get the class type
+                // https://blog.retep.org/2009/02/13/getting-class-values-from-annotations-in-an-annotationprocessor/
+                try {
+                    annotation.value();
+                }
+                catch(MirroredTypeException mte) {
+                    jsonAdapterType = mte.getTypeMirror();
+                    jsonAdapterTypeElement = TypeUtils.getElementFromTypeMirror(jsonAdapterType);
+                    TypeUtils.JsonAdapterType jsonAdapterType1 = TypeUtils.getJsonAdapterType(jsonAdapterType);
+                    for(Element element : jsonAdapterTypeElement.getEnclosedElements()) {
+                        if (element instanceof ExecutableElement) {
+                            ExecutableElement executableElement =
+                                    ((ExecutableElement) element);
+                            Name name = executableElement.getSimpleName();
+                            if (name.contentEquals("<init>")) {
+                                String fieldAdapterAccessor = getAdapterAccessorForKnownJsonAdapterType(executableElement, adapterBuilder, constructorBuilder, fieldType,
+                                        jsonAdapterType1, result, annotation.nullSafe(), getJsonName(entry.getKey()));
+                                result.addFieldToAccessor(getJsonName(entry.getKey()), fieldAdapterAccessor);
+                            }
+                        }
+                    }
+                }
```

Comment:
Could you move this annotation value grabbing code out into a utility method? It would really clean up this method. I would suggest putting it into `ElementUtils` and would ask that you write a unit test for the method that is created.